src/SA/ldapsearch/entry.c

```diff
@@ -0,0 +1,398 @@
+    LDAPMessage* pSearchResult;
+    
+	if(strlen(ldap_attributes)!=0){
+		PCHAR attr[1];
+		attr[0] = ldap_attributes;
+
+		errorCode = WLDAP32$ldap_search_s(
+        pLdapConnection,    // Session handle
+        distinguishedName,           // DN to start search
+        LDAP_SCOPE_SUBTREE, // Scope
+        ldap_filter,          // Filter
+        attr,      // Retrieve list of attributes
+        0,                  // Get both attributes and values
+        &pSearchResult);    // [out] Search results
+	} else {
+		errorCode = WLDAP32$ldap_search_s(
+        pLdapConnection,    // Session handle
+        distinguishedName,           // DN to start search
+        LDAP_SCOPE_SUBTREE, // Scope
+        ldap_filter,          // Filter
+        NULL,      // Retrieve list of attributes
+        0,                  // Get both attributes and values
+        &pSearchResult);    // [out] Search results
+	}   
+    
+    if (errorCode != LDAP_SUCCESS)
+    {
+        BeaconPrintf(CALLBACK_ERROR, "LDAP Search Failed: %i", errorCode);
+
+        WLDAP32$ldap_unbind_s(pLdapConnection);
+        if(pSearchResult != NULL)
+            WLDAP32$ldap_msgfree(pSearchResult);
+    }
+    else {
+    	internal_printf("[+] LDAP Search Succeeded.\n");
+    }
+    return pSearchResult;
+
+}
+
+void printAttribute(PCHAR pAttribute, PCHAR* ppValue){
+    ULONG iValue = 0;
+
+	// Print status if no values are returned (NULL ptr)
+    if(ppValue == NULL)
+    {
+    	internal_printf("%s: [NO ATTRIBUTE VALUE RETURNED]", pAttribute);
+    }
+
+    // Output the attribute values
+    else
+    {
+        iValue = WLDAP32$ldap_count_values(ppValue);
+        if(!iValue)
+        {
+        	internal_printf("%s: [BAD VALUE LIST]", pAttribute);
+        }
+        else
+        {
+            if((MSVCRT$strcmp(pAttribute, "pwdLastSet") == 0)||
+            	(MSVCRT$strcmp(pAttribute, "lastLogon") == 0)||
+            	(MSVCRT$strcmp(pAttribute, "badPasswordTime") == 0)||
+            	(MSVCRT$strcmp(pAttribute, "lastLogonTimestamp") == 0)){
+            	internal_printf("\n%s: ", pAttribute);
+            	// TODO: Implement transformations for various attribute data types.
+                internal_printf("%s", *ppValue);
+            	
+            } else
+                internal_printf("\n%s: %s", pAttribute, *ppValue); // Output the first attribute value
+
+            // Output more values if available
+            ULONG z;
+            for(z=1; z<iValue; z++)
+            {
+                internal_printf(", %s", ppValue[z]);
+            }
+        }
+    }
+}
+
+
+void ldapSearch(char * ldap_filter, char * ldap_attributes,	ULONG results_count){
+	ULONG numberOfEntries;
+	ULONG results_limit = 0;
+
+	internal_printf("[*] Using filter: %s\n",ldap_filter);
+	
+	if(strlen(ldap_attributes)==0)
+		internal_printf("[*] Returning all attributes.\n\n");
+	else
+		internal_printf("[*] Returning specific attribute: %s\n\n",ldap_attributes);
+	
+
+	TCHAR szDN[1024];
```

Comment:
I'd recommend against using TCHAR in general unless you also are using the various _t string handling functions, otherwise it is depending on UNICODE being either defined or not defined (not defined in this case)

In this instance it is later used with strstr so this should be a standard char []